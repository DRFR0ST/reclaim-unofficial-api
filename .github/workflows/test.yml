name: ðŸ§ª Run Tests

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  schedule:
    - cron: '0 0 * * *'  # Run every day at midnight to ensure the reverse engineered API is up to date.

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v3
    - uses: oven-sh/setup-bun@v1

    - name: Install dependencies
      run: bun install

    - name: Run tests
      uses: nick-fields/retry@v2
      with:
        timeout_minutes: 5
        max_attempts: 3
        command: bun run test
      env:
        RECLAIM_ACCESS_TOKEN: ${{ secrets.RECLAIM_ACCESS_TOKEN }}
    
    - name: Build library
      run: bun run build
      env:
        RECLAIM_ACCESS_TOKEN: ${{ secrets.RECLAIM_ACCESS_TOKEN }}