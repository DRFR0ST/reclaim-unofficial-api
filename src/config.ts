
const RECLAIM_API_URL = "https://api.app.reclaim.ai";
const RECLAIM_ACCESS_TOKEN = process?.env?.RECLAIM_ACCESS_TOKEN;
const RECLAIM_VERBOSE = process?.env?.RECLAIM_VERBOSE === "true";

/**
 * @description Configuration for the Reclaim API client.
 */
export const config = Object.freeze({
    reclaim: {
        apiUrl: RECLAIM_API_URL,
        accessToken: RECLAIM_ACCESS_TOKEN
    },
    verbose: RECLAIM_VERBOSE
});