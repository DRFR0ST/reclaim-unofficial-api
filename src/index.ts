export { ReclaimClient } from "./client"
export type { ReclaimUser, ReclaimCalendar, ReclaimTask, ReclaimEndpoints } from './interfaces';