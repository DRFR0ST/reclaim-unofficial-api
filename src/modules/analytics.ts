import {
  ReclaimEndpoints,
  ReclaimAnalytics as ReclaimAnalyticsType,
  ReclaimAnalyticsGetPayload,
  ReclaimAnalyticsMetricName,
} from "../types";
import { ReclaimClient } from "../client";

/**
 * A class for interacting with Reclaim analytics.
 */
export class ReclaimAnalytics {
  private path = ReclaimEndpoints.Analytics;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description The default metric name to use when getting analytics.
   */
  private defaultMetricName: ReclaimAnalyticsMetricName[] = [
    "DURATION_BY_CATEGORY",
    "DURATION_BY_DATE_BY_CATEGORY",
    "FOCUS_TIME_DURATION_BY_DAY_OF_WEEK",
    "MEETING_CALLOUT_GROUP",
    "MEETING_TIME_DURATION_BY_DAY_OF_WEEK",
    "PERSONAL_HABITS_AND_TASKS_DURATION_BY_DATE_BY_NAME",
    "ROI_CALLOUT_GROUP",
    "TIME_WITH_PEOPLE_DURATION_BY_PERSON",
    "WORK_HABITS_AND_TASKS_DURATION_BY_DATE_BY_NAME",
    "WORK_LIFE_CALLOUT_GROUP",
  ];

  /**
   * @description Get the user's analytics.
   * @returns {ReclaimAnalytics}
   */
  async get({
    start,
    end,
    metricName,
    groupingInterval,
  }: ReclaimAnalyticsGetPayload = {}): Promise<ReclaimAnalyticsType> {
    // Set the start and end dates to the last 7 days and the next 7 days, respectively. Format YYYY-MM-DD.
    if (!start)
      start = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000)
        .toISOString()
        .split("T")[0];
    if (!end)
      end = new Date(Date.now() + 7 * 24 * 60 * 60 * 1000)
        .toISOString()
        .split("T")[0];
    if (!metricName) metricName = this.defaultMetricName;
    if (!groupingInterval) groupingInterval = "DAILY";

    // Ensure that metricName is an array of strings
    if (!Array.isArray(metricName))
      throw new Error("The property metricName must be an array of strings.");

    // Compose the URL.
    const url = `${
      this.path
    }?start=${start}&end=${end}&metricName=${metricName.join(
      "%2C"
    )}&groupingInterval=${groupingInterval}`;

    return await this.client._fetcher(url, {
      method: "GET",
    });
  }
}
