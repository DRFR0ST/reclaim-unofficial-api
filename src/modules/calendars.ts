import {
  ReclaimCalendar,
  ReclaimEndpoints,
} from "../types";
import { ReclaimClient } from "../client";

/**
 * A class for interacting with Reclaim calendars.
 */
export class ReclaimCalendars {
  private path = ReclaimEndpoints.Calendars;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description Get the primary calendar.
   * @returns {ReclaimUser}
   */
  async primary(): Promise<ReclaimCalendar> {
    return await this.client._fetcher(`${this.path}/primary`, {
      method: "GET",
    });
  }
}
