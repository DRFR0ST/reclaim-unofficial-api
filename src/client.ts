import { ReclaimTasks } from "./modules/tasks";
import { ReclaimUsers } from "./modules/users";
import { ReclaimCalendars } from "./modules/calendars";
import { config } from "./config";

export class ReclaimClient {
  private accessToken = "";

  constructor(options?: { accessToken: string }) {
    if (!options?.accessToken && !config.reclaim.accessToken)
      throw new Error("Reclaim access token is required.");

    this.accessToken = options?.accessToken || config.reclaim.accessToken!;
  }

  get tasks() {
    return new ReclaimTasks(this);
  }

  get users() {
    return new ReclaimUsers(this);
  }

  get calendars() {
    return new ReclaimCalendars(this);
  }

  async _fetcher(endpoint: string, options: RequestInit) {
    const opts = {
      ...options,
      headers: {
        ...options.headers,
        Authorization: `Bearer ${this.accessToken}`,
      },
    };

    try {
      const response = await fetch(
        `${config.reclaim.apiUrl}/api/${endpoint}`,
        opts
      );
      const json = await response.json();

      return json;
    } catch (err) {
      console.error(err);
      throw new Error("Error fetching data from Reclaim API.");
    }
  }
}
