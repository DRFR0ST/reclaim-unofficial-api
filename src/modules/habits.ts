import {
  ReclaimEndpoints,
  ReclaimHabit,
  ReclaimHabitCreate,
} from "../types";
import { ReclaimClient } from "../client";
import { filterArray } from "../utils";

/**
 * A class for interacting with Reclaim habits.
 */
export class ReclaimHabits {
  private path = ReclaimEndpoints.Habits;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description Create a new habit.
   * @param habit The habit to create.
   * @returns {ReclaimHabit}
   */
  async create(habit: ReclaimHabitCreate): Promise<ReclaimHabit> {
    const request = await this.client._fetcher(this.path, {
      method: "POST",
      body: JSON.stringify(habit),
    });

    const createdHabit = request.find((h: ReclaimHabit) => h.title === habit.title);

    return createdHabit;
  }

  async search(filters?: Partial<ReclaimHabit>): Promise<ReclaimHabit[]> {
    const response = (await this.client._fetcher(this.path, {
      method: "GET",
    })) as ReclaimHabit[];

    if (!filters) return response;

    // Filter the response based on the filter object
    return filterArray(response, filters);
  }

  /**
   * @description Get a single habit by ID.
   * @param id The ID of the habit to get.
   * @returns {ReclaimHabit | null}
   */
  async get(id: number): Promise<ReclaimHabit | null> {
    const habits = await this.search();

    return habits.find((habit) => habit.id === id) || null;
  }

  /**
   * @description Update a habit.
   * @param habit The habit to update.
   * @returns {ReclaimHabit}
   */
  async update(id: number, habit: Partial<ReclaimHabitCreate>): Promise<ReclaimHabit> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "PATCH",
      body: JSON.stringify(habit),
    });
  }

  /**
   * @description Delete a habit.
   * @param id The ID of the habit to delete.
   * @returns {null}
   */
  async delete(id: number): Promise<{ taskOrHabit: ReclaimHabit, events: unknown[] }> {
    return await this.client._fetcher(`planner/policy/habit/${id}`, {
      method: "DELETE",
    });
  }
}
