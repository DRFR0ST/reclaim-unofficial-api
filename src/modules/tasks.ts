import {
  ReclaimEndpoints,
  ReclaimTask,
  ReclaimTaskCreate,
} from "../types";
import { ReclaimClient } from "../client";
import { filterArray } from "../utils";

/**
 * A class for interacting with Reclaim tasks.
 */
export class ReclaimTasks {
  private path = ReclaimEndpoints.Tasks;
  private client: ReclaimClient;

  constructor(client: ReclaimClient) {
    this.client = client;
  }

  /**
   * @description Get all tasks, optionally filtered by a filter object.
   * @param filters The filters to apply.
   * @returns {ReclaimTask[]}
   */
  async search(filters?: Partial<ReclaimTask>): Promise<ReclaimTask[]> {
    const response = (await this.client._fetcher(`${this.path}`, {
      method: "GET",
    })) as ReclaimTask[];

    if (!filters) return response;

    // Filter the response based on the filter object
    return filterArray(response, filters);
  }

  /**
   * @description Get a single task by ID.
   * @param id The ID of the task to get.
   * @returns {ReclaimTask}
   */
  async get(id: number): Promise<ReclaimTask> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "GET",
    });
  }

  /**
   * @description Create a new task.
   * @param task The task to create.
   * @returns {ReclaimTask}
   */
  async create(task: ReclaimTaskCreate): Promise<ReclaimTask> {
    return await this.client._fetcher(this.path, {
      method: "POST",
      body: JSON.stringify(task),
    });
  }

  /**
   * @description Update a task.
   * @param id The ID of the task to update.
   * @param task The task to update.
   * @returns {ReclaimTask}
   */
  async update(id: number, task: Partial<ReclaimTask>): Promise<ReclaimTask> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "PATCH",
      body: JSON.stringify(task),
    });
  }

  /**
   * @description Delete a task.
   * @param id The ID of the task to delete.
   * @returns {null}
   */
  async delete(id: number): Promise<null> {
    return await this.client._fetcher(`${this.path}/${id}`, {
      method: "DELETE",
    });
  }
}
