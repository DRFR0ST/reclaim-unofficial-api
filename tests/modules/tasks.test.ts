import { beforeEach, describe, expect, test } from "bun:test";
import { ReclaimClient } from "../../src";
import { ReclaimTasks } from "../../src/modules/tasks";
import { ReclaimTaskCreateMock, ReclaimTaskUpdateMock } from "../mocks";

describe("ReclaimTasks", () => {
  let tasks: ReclaimTasks;

  beforeEach(() => {
    const client = new ReclaimClient();
    tasks = client.tasks;
  });

  test("should be created", () => {
    expect(tasks).toBeTruthy();
  });

  test("should run tasks in sequence", async () => {
    // Test that tasks is created
    expect(tasks).toBeTruthy();
  
    // Test task creation
    const createResults = await tasks.create(ReclaimTaskCreateMock);
    expect(createResults).toBeTruthy();
    expect(createResults.id).toBeGreaterThan(0);
  
    // Test task search
    const searchResults = await tasks.search();
    expect(searchResults).toBeTruthy();
    expect(searchResults.length).toBeGreaterThan(0);
  
    // Test task search by title
    const filterResults = await tasks.search({ title: ReclaimTaskCreateMock.title });
    expect(filterResults).toBeTruthy();
    expect(filterResults.length).toBeGreaterThan(0);
    expect(filterResults[0].id).toBeGreaterThan(0);
    expect(filterResults[0].title).toBe(ReclaimTaskCreateMock.title);
  
    // Test task update
    const updateResults = await tasks.update(createResults.id, ReclaimTaskUpdateMock);
    expect(updateResults).toBeTruthy();
    expect(updateResults.id).toBe(updateResults.id);
    expect(updateResults.title).toBe(ReclaimTaskUpdateMock.title);
  
    // Test task get
    const getResults = await tasks.get(createResults.id);
    expect(getResults).toBeTruthy();
    expect(getResults.id).toBe(createResults.id);

    // Test task mark as done
    const markAsDoneResults = await tasks.markDone(createResults.id);
    expect(markAsDoneResults).toBeTruthy();
    expect(markAsDoneResults.taskOrHabit.id).toBe(createResults.id);
  
    // Test task delete
    const deleteResults = await tasks.delete(createResults.id);
    expect(deleteResults).toBeNull();
  });
});
