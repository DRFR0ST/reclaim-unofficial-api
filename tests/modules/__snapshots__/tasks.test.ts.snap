// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ReclaimTasks should run tasks in sequence 1`] = `
{
  "adjusted": false,
  "alwaysPrivate": true,
  "atRisk": false,
  "created": "hidden",
  "deleted": false,
  "due": "2029-11-21T17:30:00+01:00",
  "eventCategory": "WORK",
  "eventSubType": "FOCUS",
  "id": "hidden",
  "index": "hidden",
  "maxChunkSize": 8,
  "minChunkSize": 4,
  "notes": "",
  "onDeck": false,
  "priority": "P2",
  "readOnlyFields": [],
  "recurringAssignmentType": "TASK",
  "snoozeUntil": "2029-11-17T07:00:00+01:00",
  "sortKey": "hidden",
  "status": "NEW",
  "taskSource": {
    "type": "RECLAIM_APP",
  },
  "timeChunksRemaining": 8,
  "timeChunksRequired": 8,
  "timeChunksSpent": 0,
  "timeSchemeId": "989b3027-46c4-4729-bdec-1070fc4d8c0f",
  "title": "Funky Imitation Game",
  "type": "TASK",
  "updated": "hidden",
}
`;
